lib/charms/airflow_coordinator_k8s/v0/airflow_coordinator_temp.py

```diff
@@ -0,0 +1,836 @@
+            )
+
+            if config_template:
+                model.config_template = config_template
+
+            if kubernetes_executor_pod_spec:
+                model.kubernetes_executor_pod_spec = kubernetes_executor_pod_spec
+
+            if sensitive_data:
+                model.sensitive_data = json.dumps(sensitive_data)
+        except pydantic.ValidationError:
+            model = AirflowCoordinatorProviderModel(
+                config_template=config_template,
+                kubernetes_executor_pod_spec=kubernetes_executor_pod_spec,
+                sensitive_data=json.dumps(sensitive_data),
+            )
+
+        for relation in self.interface.relations:
+            self.interface.write_model(relation.id, model)
+
+    def set_validation_errors(self, failures: list[MetadataValidationError]) -> None:
+        """Update validation errors to send to related core charms."""
+        if not self.interface.relations:
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        failures_serialized = json.dumps([failure.model_dump() for failure in failures])
+
+        try:
+            model = self.interface.build_model(self.relation.id, AirflowCoordinatorProviderModel)
+            model.validation_failures = failures_serialized
+        except pydantic.ValidationError:
+            model = AirflowCoordinatorProviderModel(
+                validation_failures=failures_serialized,
+            )
+
+        for relation in self.interface.relations:
+            self.interface.write_model(relation.id, model)
+
+    @property
+    def core_charms_metadata(self) -> dict[str, AirflowCoordinatorRequirerModel]:
+        """Charm metadata from each of the related core charms."""
+
+        def _build_requirer_model(
+            relation: ops.Relation,
+        ) -> typing.Optional[AirflowCoordinatorRequirerModel]:
+            try:
+                return self.interface.build_model(
+                    relation.id, AirflowCoordinatorRequirerModel, component=relation.app
+                )
+            except pydantic.ValidationError:
+                return None
+
+        return {
+            metadata.component: metadata
+            for metadata in [
+                _build_requirer_model(relation)
+                for relation in self.interface.relations
+                if self.interface.repository(relation.id, relation.app).get_data()
+            ]
+            if metadata is not None
+        }
+
+
+class AirflowCoordinatorRequires(ops.Object):
+    """A requirer handler encapsulating the airflow coordinator relation."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        relation_name: str,
+        component: str,
+        workload_container_name: str,
+        callback: typing.Callable,
+    ):
+        if component not in REQUIRED_AIRFLOW_CORE_COMPONENTS:
+            raise ValueError(f"Invalid component {component}")
```

Comment:
If we already have the `AirflowCoordinatorRequirerModel`, would it make sense to let pydantic do this check?


```suggestion
        AirflowCoordinatorRequirerModel(
            airflow_version="3.1"
            workload_image_hash="somehash",
            component=component, # <-- check the required component
        )
```

Reply:
simplified all validators to eliminate redundant validations + using enums where it makes sense (9fee0f46b1e599a7112889db01941e444cdc84cf)